src/intrinsic/mod.rs

```diff
@@ -819,75 +820,35 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
-
-        // First step.
-        let mask = self.context.new_rvalue_from_long(value_type, 0x5555555555555555);
-        let left = value & mask;
-        let shifted = value >> self.context.new_rvalue_from_int(value_type, 1);
-        let right = shifted & mask;
-        let value = left + right;
-
-        // Second step.
-        let mask = self.context.new_rvalue_from_long(value_type, 0x3333333333333333);
-        let left = value & mask;
-        let shifted = value >> self.context.new_rvalue_from_int(value_type, 2);
-        let right = shifted & mask;
-        let value = left + right;
-
-        // Third step.
-        let mask = self.context.new_rvalue_from_long(value_type, 0x0F0F0F0F0F0F0F0F);
-        let left = value & mask;
-        let shifted = value >> self.context.new_rvalue_from_int(value_type, 4);
-        let right = shifted & mask;
-        let value = left + right;
-
-        if value_type.is_u8(&self.cx) {
-            return self.context.new_cast(None, value, result_type);
-        }
-
-        // Fourth step.
-        let mask = self.context.new_rvalue_from_long(value_type, 0x00FF00FF00FF00FF);
-        let left = value & mask;
-        let shifted = value >> self.context.new_rvalue_from_int(value_type, 8);
-        let right = shifted & mask;
-        let value = left + right;
-
-        if value_type.is_u16(&self.cx) {
-            return self.context.new_cast(None, value, result_type);
-        }
-
-        // Fifth step.
-        let mask = self.context.new_rvalue_from_long(value_type, 0x0000FFFF0000FFFF);
-        let left = value & mask;
-        let shifted = value >> self.context.new_rvalue_from_int(value_type, 16);
-        let right = shifted & mask;
-        let value = left + right;
-
-        if value_type.is_u32(&self.cx) {
-            return self.context.new_cast(None, value, result_type);
-        }
-
-        // Sixth step.
-        let mask = self.context.new_rvalue_from_long(value_type, 0x00000000FFFFFFFF);
-        let left = value & mask;
-        let shifted = value >> self.context.new_rvalue_from_int(value_type, 32);
-        let right = shifted & mask;
-        let value = left + right;
+        let (popcount, expected_type) =
+            // TODO(antoyo): write a new function Type::is_compatible_with(&Type) and use it here
+            // instead of using is_uint().
+            if arg_type.is_uchar(&self.cx) || arg_type.is_ushort(&self.cx) || arg_type.is_uint(&self.cx) {
+                ("__builtin_popcount", self.cx.uint_type)
+            }
+            else if arg_type.is_ulong(&self.cx) {
+                ("__builtin_popcountl", self.cx.ulong_type)
+            }
+            else if arg_type.is_ulonglong(&self.cx) {
+                ("__builtin_popcountll", self.cx.ulonglong_type)
+            } else {
+                // TODO(antoyo): implement in the normal algorithm below to have a more efficient
+                // implementation (that does not require a call to __popcountdi2).
```

Comment:
I wrote a comment here that explains why I didn't use that popcount builtins.
I don't quite recall, but I believe I had some undefined references to some `__popcount*` function in some cases.
Do you have any idea when this could happen?

Perhaps that was in `#[no_core]`.

Reply:
As far as I understand, you might run into it if you don't have the compiler runtime (e.g. `libgcc_s.so.1`, compiler-rt, etc.) available.  They get linked in for a variety of reasons (the actual link directives happen in [unwind/src/lib.rs][1]), though I think you can disable linking to them if you have the `crt-static` target feature set during the build process.

[1]:  https://github.com/rust-lang/rust/blob/cae0791da47bb01f16885eb233dcd66b0093a6e1/library/unwind/src/lib.rs

* * * 

To be honest, I misread that comment - I thought it was referring to the impl for 128-bit ints, not the rest of the impl.  My mistake.

Reply:
Maybe the solution here is to call the built-ins unless `crt-static` is set, in which case we fall back on an inline implemenatation of our own.

Reply:
Actually, I think this discussion might be moot for now.  By my own testing, gcc seems to have better pattern matching for Wenger's algorithm[^1] - while it's not 100% perfect, it seems to be optimized a lot better than the current impl (at least on x86_64, which is all I can test on at the moment) and it solves the problem of generating library calls to `libgcc_s.so`.

I've adjusted my implementation to reflect this new approach in 5ae3d85.

[^1]: Discussion for Wenger's algorithm can be found [here](https://en.wikipedia.org/wiki/Hamming_distance); the original paper may be found [here](https://dl.acm.org/doi/pdf/10.1145/367236.367286).